pkgs/ok_http/android/src/main/kotlin/com/example/ok_http/RedirectInterceptor.kt

```diff
@@ -11,8 +11,19 @@ package com.example.ok_http
 
 import okhttp3.Interceptor
 import okhttp3.OkHttpClient
+import okhttp3.Response
 import java.io.IOException
 
+/**
+ * Callback interface utilized by the [RedirectInterceptor].
+ *
+ * Allows Dart code to operate upon the intermediate redirect responses.
+ */
+interface RedirectReceivedCallback {
+    fun onRedirectReceived(response: Response, location: String)
```

Comment:
In a follow-up PR, would it make sense to migrate the redirect logic to Dart? Actually, would it make sense to make interceptors available in Dart?

Taking a step back, maybe you'd want to expose `OkHttpClient` or a wrapper around that to Dart?

With that users can control caching, connection pools, DNS, etc.

I'm not sure what the best way to do that is though.

Reply:
About migrating the redirection logic to Dart, there are issues with throwing exceptions when the interface (`Interceptor`) that we're implementing, is being consumed on another thread.
Hossein and I have been trying to work our way around this issue, starting with https://github.com/dart-lang/native/pull/1211. But so far, `RedirectInterceptor` stands out as an example of this drawback.

---

About exposing configurations to users, I had planned to provide high-level functions (and constructor parameters) within `OkHttpClient` allowing users to tweak the client according to their requirements.

For example, if a user wants to alter the timeout for a request, they can either pass the timeout duration in the constructor or in the function:
```dart
class OkHttpClient extends BaseClient {
  late bindings.OkHttpClient _client;
  OkHttpClient([Duration? timeoutDuration]) {
    // creates a client based on the timeout and assigns it to _client
  }
  void setTimeout(Duration timeoutDuration) {
    // uses a new builder from `_client` and edits its timeout
  }
```

But yes, a wrapper sounds good! We can use the architecture that `CronetEngine` has.